### PR TITLE
Fix ICE #3717 in lint implicit_hasher

### LINKED ISSUE #3717: ICE in https://github.com/remram44/rs-sync
````
git clone https://github.com/remram44/rs-sync
git checkout 452a2351a039fdfc68e3386187ae1a57620687a4
RUSTC_BACKTRACE=Full cargo clippy -j 1
````

````
   --> src/delta.rs:209:47
    |
209 |                 if try!(file.read(&mut buffer[idx..idx + 1])) == 0 {
    |                                               ^^^^^^^^^^^^ help: use: `idx..=idx`
    |
    = note: #[warn(clippy::range_plus_one)] on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#range_plus_one

error: internal compiler error: src/librustc/ty/context.rs:528: node_id_to_type: no type for node `expr <io::Error>::new(io::ErrorKind::InvalidData, "Invalid delta file") (id=2218)`

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:589:9
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
error: aborting due to previous error


note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.33.0-nightly (bf669d1e3 2019-01-25) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental -C target-cpu=native --crate-type lib

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `rs-sync`.

To learn more, run the command again with --verbose.
````

`clippy 0.0.212 (f175352 2019-01-24)`

### PR DESCRIPTION
**Pull request by flip1995**
Fixes #3717 

This fixes the ICE. We lose some information in a very specific case though. But less information if better than an ICE. For an example see the test file.

Does anyone know, if there's another way to get the `ty::Ty` of a `hir::Expr`?

### CODE REVIEW

File: clippy_lints/src/types.rs [hunk 1]
```diff
@@ -2248,11 +2248,14 @@ impl<'a, 'b, 'tcx: 'a + 'b> Visitor<'tcx> for ImplicitHasherConstructorVisitor<'
             if let ExprKind::Call(ref fun, ref args) = e.node;
             if let ExprKind::Path(QPath::TypeRelative(ref ty, ref method)) = fun.node;
             if let TyKind::Path(QPath::Resolved(None, ref ty_path)) = ty.node;
+            // ICE #3717; If the TypechTables don't have information about the expression, we can't
```
Comment (oli-obk) [inline clippy_lints/src/types.rs]:
```suggestion
            // ICE #3717; If the TypeckTables don't have information about the expression, we can't
```

File: tests/ui/ice-3717.rs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+use std::collections::HashSet;
+
+fn main() {}
+
+pub fn ice_3717(_: &HashSet<usize>) {
+    let _ = [0u8; 0];
+    // FIXME: This should produce a suggestion, but we need to bail out here, because
+    // `TypeckTables::expr_ty(..)` panics when we bind an `ExprKind::Repeat(..)` before any `new()`
```
Comment (oli-obk) [inline tests/ui/ice-3717.rs]: lol wat? What's going on here? :laughing: 

Are we maybe overwriting our TypeckTables when going into the `0` constant and not resetting it after leaving said constant? I can't think of a way that the `TypeckTables` would get corrupted in rustc.

Comment (oli-obk) [reply tests/ui/ice-3717.rs]: Ah indeed. in `visit_body` we only set the new body, and don't reset it to the previous body afterwards.

Comment (flip1995) [reply tests/ui/ice-3717.rs]: Ahh, good catch. Fixed it!

File: clippy_lints/src/types.rs [hunk 3]
```diff
@@ -2239,20 +2239,19 @@ impl<'a, 'b, 'tcx: 'a + 'b> ImplicitHasherConstructorVisitor<'a, 'b, 'tcx> {
...
         walk_body(self, body);
+        self.body = prev_body;
     }
 
     fn visit_expr(&mut self, e: &'tcx Expr) {
         if_chain! {
             if let ExprKind::Call(ref fun, ref args) = e.node;
             if let ExprKind::Path(QPath::TypeRelative(ref ty, ref method)) = fun.node;
             if let TyKind::Path(QPath::Resolved(None, ref ty_path)) = ty.node;
+            if same_tys(self.cx, self.target.ty(), self.body.expr_ty(e));
```
Comment (oli-obk) [inline clippy_lints/src/types.rs]: this doesn't quite do the same thing. we used to not walk the expression if this check failed. Not sure what for though. Maybe performance?

Comment (flip1995) [reply clippy_lints/src/types.rs]: Yeah this should only differ in performance, I'll revert this change

**Comment by flip1995** [comment]
Done! Also squashed the commits.

**Pull request merged**